secp256kfun/src/poly.rs

```diff
@@ -0,0 +1,260 @@
+///
+/// # Example
+///
+/// ```
+/// use secp256kfun::{g, poly, s, Scalar, G};
+/// let secret_poly = (0..5)
+///     .map(|_| Scalar::random(&mut rand::thread_rng()))
+///     .collect::<Vec<_>>();
+/// let point_poly = poly::to_point_poly(&secret_poly);
+/// ```
+pub fn to_point_poly(scalar_poly: &[Scalar]) -> Vec<Point> {
+    scalar_poly.iter().map(|a| g!(a * G).normalize()).collect()
+}
+
+/// Generate a [`Scalar`] polynomial for key generation
+///
+/// [`Scalar`]: crate::Scalar
+pub fn generate_scalar_poly(threshold: usize, rng: &mut impl RngCore) -> Vec<Scalar> {
+    (0..threshold).map(|_| Scalar::random(rng)).collect()
+}
+
+/// Evaluate a scalar polynomial defined by coefficients, at some scalar index.
+///
+/// The polynomial coefficients begin with the smallest degree term first (the constant).
+pub fn scalar_poly_eval(
+    poly: &[Scalar],
+    x: Scalar<impl Secrecy, impl ZeroChoice>,
+) -> Scalar<Secret, Zero> {
+    poly.iter()
+        .fold((s!(0), s!(1).mark_zero()), |(eval, xpow), coeff| {
+            (s!(eval + xpow * coeff), s!(xpow * x))
+        })
+        .0
+}
+
+/// Evaluate a point polynomial defined by coefficients, at some scalar index.
+///
+/// The polynomial coefficients begin with the smallest degree term first (the constant).
+pub fn point_poly_eval(
+    poly: &[Point<impl PointType, Public, impl ZeroChoice>],
+    x: Scalar<Public, impl ZeroChoice>,
+) -> Point<NonNormal, Public, Zero> {
+    let xpows = core::iter::successors(Some(s!(1).public().mark_zero()), |xpow| {
+        Some(s!(xpow * x).public())
+    })
+    .take(poly.len())
+    .collect::<Vec<_>>();
+    crate::op::lincomb(&xpows, poly.iter())
+}
+
+/// Calculate the lagrange coefficient for participant with index x_j and other signers indexes x_ms
+pub fn lagrange_lambda(
+    x_j: Scalar<impl Secrecy>,
+    x_ms: impl Iterator<Item = Scalar<impl Secrecy>>,
+) -> Scalar<Public> {
+    x_ms.fold(Scalar::one(), |acc, x_m| {
+        let denominator = s!(x_m - x_j)
+            .non_zero()
+            .expect("indexes must be unique")
+            .invert();
+        s!(acc * x_m * denominator).public()
+    })
+}
+
+/// Find the coefficients of the polynomial that interpolates a set of points at unique indexes.
+///
+/// Panics if the indexes are not unique.
+///
+/// A vector with a tail of zero coefficients means the interpolation was overdetermined.
+pub fn interpolate_point_polynomial(
+    indexes: Vec<Scalar<impl Secrecy, impl ZeroChoice>>,
+    points: Vec<Point>,
+) -> Vec<Point<impl PointType, Public, Zero>> {
+    // Get each lagrange basis polynomial from the product of coefficients:
+    //      l_j(x) = Product[ (x-x_m)/(x_j-x_m), j!=m]
+    // Or
+    //      l_j(x) = Product[ a_m*x + b_m, j!=m], where a_m = 1/(x_j-x_m) and b_m = -x_m*a_m.
+    let basis_polynomials: Vec<_> = indexes
+        .clone()
+        .into_iter()
+        .enumerate()
+        .map(|(j, x_j)| {
```

Comment:
I think this bit of the map could go into its own function like:

```rust
fn unit_basis_poly(indices: &[Scalar<Public, impl ZeroChoice>] ) -> Vec<Vec<Scalar<Public, Zero>>>
```
I call this the unit basis because the basis is the one after you multiply by `y` I think. There's probably a better term for this. Not sure if it's a good idea.